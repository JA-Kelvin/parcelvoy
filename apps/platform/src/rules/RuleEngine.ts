import { TemplateEvent } from '../users/UserEvent'
import { TemplateUser, User } from '../users/User'
import Rule, { AnyJson, RuleTree, Operator, RuleGroup, RuleType } from './Rule'
import NumberRule from './NumberRule'
import StringRule from './StringRule'
import BooleanRule from './BooleanRule'
import DateRule from './DateRule'
import ArrayRule from './ArrayRule'
import WrapperRule from './WrapperRule'
import { uuid } from '../utilities'
import App from '../app'

class Registry<T> {
    #registered: { [key: string]: T } = {}

    public register(key: string, obj: T) {
        this.#registered[key] = obj
        return this
    }

    public get(key: string) {
        return this.#registered[key]
    }
}

export interface RuleCheckInput {
    user: TemplateUser
    events: TemplateEvent[] // all of this user's events
}

export interface RuleBaseParams {
    registry: typeof ruleRegistry
    rule: RuleTree // current rule to use
}

export interface RuleQueryParams extends RuleBaseParams {
    projectId: number
}

export interface RuleCheckParams extends RuleBaseParams {
    input: RuleCheckInput // all contextual input data
    value: Record<string, unknown> // current value to evaluate against
}

export interface RuleCheck {
    check(params: RuleCheckParams): boolean
    query(params: RuleQueryParams): string
}

const ruleRegistry = new Registry<RuleCheck>()

export class RuleEvalException extends Error {
    constructor(rule: Rule | RuleTree, message: string) {
        super(message)
    }
}

ruleRegistry.register('number', NumberRule)
ruleRegistry.register('string', StringRule)
ruleRegistry.register('boolean', BooleanRule)
ruleRegistry.register('date', DateRule)
ruleRegistry.register('array', ArrayRule)
ruleRegistry.register('wrapper', WrapperRule)

export const check = (input: RuleCheckInput, rule: RuleTree | RuleTree[]) => {
    if (Array.isArray(rule)) {
        rule = make({
            type: 'wrapper',
            operator: 'and',
            children: rule,
        })
    }
    return ruleRegistry.get(rule.type).check({ registry: ruleRegistry, input, rule, value: input.user })
}

export const checkQuery = async (user: User, rule: RuleTree | RuleTree[]) => {
    const subquery = getRuleQuery(user.project_id, rule)
    const query = `select exists(${subquery}) as check`
    const result = await App.main.clickhouse.query({
        query,
        format: 'JSONEachRow',
    })
    const data = await result.json() as { check: boolean }[]
    return data[0].check
}

export const getRuleQuery = (projectId: number, rule: RuleTree | RuleTree[]) => {
    if (Array.isArray(rule)) {
        rule = make({
            type: 'wrapper',
            operator: 'and',
            children: rule,
        })
    }
    return ruleRegistry.get(rule.type).query({ projectId, registry: ruleRegistry, rule })
}

interface RuleMake {
    type: RuleType
    group?: RuleGroup
    path?: string
    operator?: Operator
    value?: AnyJson
    children?: RuleTree[]
}

export const make = ({ type, group = 'user', path = '$', operator = '=', value, children }: RuleMake): RuleTree => {
    const rule = {
        uuid: uuid(),
        type,
        group,
        path,
        operator,
        value,
        children,
    }

    children?.forEach(child => {
        child.parent_uuid = rule.uuid
    })

    return rule
}
