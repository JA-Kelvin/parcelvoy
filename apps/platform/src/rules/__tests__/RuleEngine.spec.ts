import { subDays } from 'date-fns'
import { check, make, getRuleQuery } from '../RuleEngine'

describe('RuleEngine', () => {
    describe('wrapper', () => {
        describe('event name match', () => {
            const setup = () => {
                const id = 'abcd'
                const name = 'Account Created'
                const rule = make({
                    type: 'wrapper',
                    operator: 'and',
                    children: [
                        // match user id
                        make({
                            type: 'string',
                            operator: '=',
                            path: 'id',
                            value: id,
                        }),
                        // this should make if the user has done this event ever
                        make({
                            type: 'wrapper',
                            group: 'event',
                            path: 'name',
                            value: name,
                            operator: 'and',
                        }),
                    ],
                })
                return { id, name, rule }
            }

            test('check', () => {
                const { id, name, rule } = setup()
                const input = {
                    user: {
                        id,
                    },
                    events: [
                        {
                            name,
                        },
                    ],
                }
                const shouldPass = check(input, rule)
                expect(shouldPass).toBeTruthy()
            })

            test('query', () => {
                const { rule } = setup()
                const result = getRuleQuery(1, rule)
                expect(result).toMatchSnapshot()
            })
        })
    })

    describe('string', () => {
        describe('equals', () => {
            test('check', () => {
                const email = 'test@test.com'
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email,
                            name: 'Name',
                        },
                        events: [],
                    },
                    make({
                        type: 'string',
                        operator: '=',
                        path: 'email',
                        value: email,
                    }),
                )
                expect(shouldPass).toBeTruthy()
            })

            test('query', () => {
                const email = 'test@test.com'
                const query = getRuleQuery(
                    1,
                    make({
                        type: 'string',
                        operator: '=',
                        path: 'email',
                        value: email,
                    }),
                )
                expect(query).toMatchSnapshot()
            })
        })

        test('does not equals', () => {
            const email = 'test@test.com'
            const shouldPass = check(
                {
                    user: {
                        id: 'abcd',
                        email,
                        name: 'Name',
                    },
                    events: [],
                },
                make({ type: 'string', path: '$.email', operator: '!=', value: email }),
            )
            expect(shouldPass).toBeFalsy()
        })

        describe('is set', () => {
            test('check', () => {
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                        },
                        events: [],
                    },
                    make({ type: 'string', path: '$.project', operator: 'is set' }),
                )
                expect(shouldPass).toBeFalsy()
            })

            test('query', () => {
                const query = getRuleQuery(
                    1,
                    make({ type: 'string', path: '$.project', operator: 'is set' }),
                )
                expect(query).toMatchSnapshot()
            })
        })
    })

    describe('date', () => {
        test('is set', () => {
            const shouldPass = check(
                {
                    user: {
                        id: 'abcd',
                        email: 'test@test.com',
                        createdAt: Date.now(),
                    },
                    events: [],
                },
                make({ type: 'date', path: '$.createdAt', operator: 'is set' }),
            )
            expect(shouldPass).toBeTruthy()
        })

        test('greater than or equals', () => {
            const now = Date.now()
            const shouldPass = check(
                {
                    user: {
                        id: 'abcd',
                        email: 'test@test.com',
                        createdAt: now,
                    },
                    events: [],
                },
                make({ type: 'date', path: '$.createdAt', operator: '>=', value: now }),
            )
            expect(shouldPass).toBeTruthy()
        })

        test('compilation', () => {
            const shouldPass = check(
                {
                    user: {
                        id: 'abcd',
                        email: 'test@test.com',
                        createdAt: subDays(new Date(), 1).getTime(),
                    },
                    events: [],
                },
                make({ type: 'date', path: '$.createdAt', operator: '>', value: '{{subDate "now" 1 "months" }}' }),
            )
            expect(shouldPass).toBeTruthy()
        })
    })

    describe('multiple', () => {
        describe('combination event and user and types', () => {
            const rules = [
                make({ type: 'string', path: '$.project', operator: 'is set' }),
                make({
                    type: 'wrapper',
                    group: 'event',
                    path: 'name',
                    value: 'beat-game',
                    operator: 'and',
                    children: [
                        make({
                            type: 'number',
                            path: 'score.total',
                            operator: '<=',
                            value: '5',
                        }),
                        make({
                            type: 'boolean',
                            path: 'score.isRecord',
                        }),
                    ],
                }),
            ]

            test('check', () => {
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                            project: 'Parcelvoy',
                        },
                        events: [
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                        ],
                    },
                    rules,
                )
                expect(shouldPass).toBeTruthy()
            })

            test('query', () => {
                const result = getRuleQuery(1, rules)
                expect(result).toMatchSnapshot()
            })
        })

        describe('combination of conditional clauses on rules', () => {
            const rules = [
                make({ type: 'string', path: '$.project', operator: 'is set' }),
                make({
                    group: 'event',
                    path: 'name',
                    value: 'beat-game',
                    type: 'wrapper',
                    operator: 'or',
                    children: [
                        make({ type: 'number', path: 'score.total', operator: '<', value: 5 }),
                        make({ type: 'boolean', path: 'score.isRecord', value: true }),
                    ],
                }),
            ]

            test('check', () => {
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                            project: 'Parcelvoy',
                        },
                        events: [
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                        ],
                    },
                    rules,
                )
                expect(shouldPass).toBeTruthy()
            })

            test('query', () => {
                const result = getRuleQuery(1, rules)
                expect(result).toMatchSnapshot()
            })
        })
    })

    describe('time bound event rules', () => {
        describe('check', () => {
            const rules = [
                make({
                    group: 'event',
                    path: 'name',
                    value: 'beat-game',
                    type: 'wrapper',
                    operator: 'or',
                    children: [
                        make({ type: 'number', path: 'score.total', operator: '<', value: 5 }),
                        make({ type: 'boolean', path: 'score.isRecord', value: true }),
                    ],
                    frequency: {
                        period: {
                            unit: 'day',
                            value: 7,
                            type: 'rolling',
                        },
                        count: 2,
                        operator: '>=',
                    },
                }),
            ]

            test('if occured more than number of times in period true', () => {
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                            project: 'Parcelvoy',
                        },
                        events: [
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                        ],
                    },
                    rules,
                )
                expect(shouldPass).toBeTruthy()
            })

            test('if occured less than number of times in period false', () => {
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                            project: 'Parcelvoy',
                        },
                        events: [
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                        ],
                    },
                    rules,
                )
                expect(shouldPass).toBeFalsy()
            })

            test('if operator is less than ensure we dont break early', () => {
                const rules = [
                    make({
                        group: 'event',
                        path: 'name',
                        value: 'beat-game',
                        type: 'wrapper',
                        operator: 'or',
                        children: [
                            make({ type: 'number', path: 'score.total', operator: '<', value: 5 }),
                            make({ type: 'boolean', path: 'score.isRecord', value: true }),
                        ],
                        frequency: {
                            period: {
                                unit: 'day',
                                value: 7,
                                type: 'rolling',
                            },
                            count: 1,
                            operator: '<=',
                        },
                    }),
                ]
                const shouldPass = check(
                    {
                        user: {
                            id: 'abcd',
                            email: 'test@test.com',
                            name: 'Name',
                            project: 'Parcelvoy',
                        },
                        events: [
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                            {
                                name: 'beat-game',
                                score: {
                                    total: 5,
                                    isRecord: true,
                                },
                            },
                        ],
                    },
                    rules,
                )
                expect(shouldPass).toBeFalsy()
            })
        })
    })
})
