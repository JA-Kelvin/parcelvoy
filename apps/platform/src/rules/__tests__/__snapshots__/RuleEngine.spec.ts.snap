// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleEngine multiple combination event and user and types query 1`] = `
"SELECT id FROM users FINAL PREWHERE \`data\`.\`project\` IS NOT NULL WHERE project_id = 1 INTERSECT 
        SELECT user_id AS id 
        FROM user_events 
        WHERE project_id = 1 and name = 'beat-game' and (accurateCastOrNull(\`data\`.\`score\`.\`total\`, 'UInt8') <= 5 and \`data\`.\`score\`.\`isRecord\` = false)
        GROUP BY project_id, user_id
        HAVING count() >= 1"
`;

exports[`RuleEngine multiple combination of conditional clauses on rules query 1`] = `
"SELECT id FROM users FINAL PREWHERE \`data\`.\`project\` IS NOT NULL WHERE project_id = 1 INTERSECT 
        SELECT user_id AS id 
        FROM user_events 
        WHERE project_id = 1 and name = 'beat-game' and (accurateCastOrNull(\`data\`.\`score\`.\`total\`, 'UInt8') < 5 or \`data\`.\`score\`.\`isRecord\` = true)
        GROUP BY project_id, user_id
        HAVING count() >= 1"
`;

exports[`RuleEngine string equals query 1`] = `"accurateCastOrNull(email, 'String') = 'test@test.com'"`;

exports[`RuleEngine string is set query 1`] = `"\`data\`.\`project\` IS NOT NULL"`;

exports[`RuleEngine wrapper event name match query 1`] = `
"SELECT id FROM users FINAL PREWHERE accurateCastOrNull(\`data\`.\`id\`, 'String') = 'abcd' WHERE project_id = 1 INTERSECT 
        SELECT user_id AS id 
        FROM user_events 
        WHERE project_id = 1 and name = 'Account Created'
        GROUP BY project_id, user_id
        HAVING count() >= 1"
`;
